ext/node/errors.rs

```diff
@@ -1,170 +1,462 @@
 // Copyright 2018-2024 the Deno authors. All rights reserved. MIT license.
 
+use std::borrow::Cow;
 use std::path::PathBuf;
 
 use deno_core::error::generic_error;
-use deno_core::error::type_error;
 use deno_core::error::AnyError;
-use deno_core::url::Url;
+use deno_core::ModuleSpecifier;
+use thiserror::Error;
 
+use crate::NodeModuleKind;
 use crate::NodeResolutionMode;
 
-pub fn err_invalid_module_specifier(
-  request: &str,
-  reason: &str,
-  maybe_base: Option<String>,
-) -> AnyError {
-  let mut msg = format!(
-    "[ERR_INVALID_MODULE_SPECIFIER] Invalid module \"{request}\" {reason}"
-  );
+macro_rules! kinded_err {
+  ($name:ident, $kind_name:ident) => {
+    #[derive(Error, Debug)]
+    #[error(transparent)]
+    pub struct $name(pub Box<$kind_name>);
 
-  if let Some(base) = maybe_base {
-    msg = format!("{msg} imported from {base}");
-  }
+    impl $name {
+      pub fn as_kind(&self) -> &$kind_name {
+        &self.0
+      }
 
-  type_error(msg)
+      pub fn into_kind(self) -> $kind_name {
+        *self.0
+      }
+    }
+
+    impl<E> From<E> for $name
+    where
+      $kind_name: From<E>,
+    {
+      fn from(err: E) -> Self {
+        $name(Box::new($kind_name::from(err)))
+      }
+    }
+  };
 }
```

Comment:
Not sure the best way to go about doing this, but I wanted all the error types to be boxed without much trouble (similar to this https://github.com/dtolnay/thiserror/issues/57#issuecomment-576437396), but still have the ability to inspect the errors. Using a macro here helped with this, but I feel like there must be something in the rust ecosystem that makes this easier.